docs/source/cuda/examples.rst

```diff
@@ -3,6 +3,479 @@
+               is_sess_boundary = 1
+         else:
+               is_sess_boundary = 0
+      else:
+         is_sess_boundary = 1
+         
+         
+      ### DETERMINE SESSION LABELS
+      # this thread marks the start of a session
+      if gid < size:
+         if is_sess_boundary:
+               results[gid] = gid
+               look_ahead = 1
+
+               # check elements 'forward' of this one 
+               # until a new session boundary is found
+               while results[gid + look_ahead] == 0:
+                  results[gid + look_ahead] = gid
+                  look_ahead += 1
+
+                  # don't segfault if I'm the last thread
+                  if gid + look_ahead == size - 1:
+                     results[gid + look_ahead] = gid
+                     break
+
+Let's generate some data and try out the kernel.
+
+.. code-block:: python
+
+   # generate data
+   ids = cuda.to_device(
+      np.array([1,1,1,1,1,1,2,2,2,3,3,3,3,3,3,3,3,3,3,4,4,4,4,4,4,4,4,4])
+   )
+
+   sec = cuda.to_device(
+      np.array(
+         [1,2,3,5000,5001,5002,1,2,3,1,2,5000,5001,10000,10001,10002,10003,15000,150001,1,5000,50001,15000,20000,25000,25001,25002, 25003],
+         dtype='datetime64[ns]').astype('int64') # cast to int64 for compatibility
+   )
+
+   # create a vector to hold the results
+   results = cuda.to_device(np.zeros(len(ids)))
+
+   # kernel call
+   sessionize.forall(len(ids))(ids, sec, results, len(ids))
+   
+   print(results.copy_to_host())
+   # array([ 0.,  0.,  0.,  3.,  3.,  3.,  6.,  6.,  6.,  9.,  9., 11., 11.,
+   #    13., 13., 13., 13., 17., 18., 19., 20., 21., 21., 23., 24., 24.,
+   #    24., 24.])
+
+
+As can be seen above, the kernel successfully divided the first three datapoints from the second three for the first user ID,
+and a similar pattern is seen througout. This example has wide applicability in a business setting and can be used to study
+many problems around consumer behavior and product interaction.
+
+.. _cuda_reuse_function:
+
+JIT Function CPU-GPU Compatibility
+==================================
+
+This example demonstrates how ``numba.jit`` can be used to jit compile a function for the CPU, while at the same time making 
+it available for use inside CUDA kernels. This can be very useful for users that are migrating workflows from CPU to GPU as 
+they can directly reuse potential business logic with fewer code changes.
+
+Take the following example function:
+
+.. code-block:: python
+
+   from math import pi
+
+   @numba.jit
+   def business_logic(x, y, z):
+      return 4 * z * (2 * x - (4 * y) / 2 * pi)
+
+The function ``business_logic`` can be run standalone in compiled form on the CPU:
+
+..code-block:: python
+
+   print(business_logic(1,2,3)) # -126.79644737231007
+
+It can also be directly reused threadwise inside a GPU kernel. For example one may 
+generate some vectors to represent ``x``, ``y``, and ``z``:
+
+.. code-block:: python
+
+   X = cp.array([1, 10, 234])
```

Comment:
Although I do love to champion CuPy / Numba interoperability, we should probably keep the Numba examples self-contained and use Numba device arrays here.